File: script/graphics/gui/frame.lua
local ffi = require "ffi"
local base = require "common.base"

require "graphics.gui.types"
local text = require "graphics.gui.text"
local texture = require "graphics.texture"

ffi.cdef [[

typedef struct ilG_gui_frame ilG_gui_frame;

enum ilG_gui_inputaction {
    ILG_GUI_UNHANDLED,
    ILG_GUI_PASSTHROUGH,
    ILG_GUI_OVERRIDE
};

typedef enum ilG_gui_inputaction (*ilG_gui_onClick)(ilG_gui_frame *self, int x, int y, int button);
typedef enum ilG_gui_inputaction (*ilG_gui_onHover)(ilG_gui_frame *self, int x, int y);
typedef void (*ilG_gui_draw_fn)(ilG_gui_frame *self, ilG_gui_rect where);

struct ilG_gui_frame {
    il_base base;
    ilG_gui_frame *parent;
    ilG_gui_rect rect;
    ilG_gui_onClick click;
    ilG_gui_onHover hover;
    ilG_gui_draw_fn draw;
    struct {
        ilG_gui_frame** data;
        size_t length;
        size_t capacity;
    } children;
    struct ilG_context *context;
};

extern il_type ilG_gui_frame_type;

void ilG_gui_frame_filler(ilG_gui_frame *self, float col[4]);
void ilG_gui_frame_image(ilG_gui_frame *self, struct ilG_texture *tex, int premultiplied);
ilG_gui_rect ilG_gui_frame_abs(ilG_gui_frame *self);
int ilG_gui_frame_contains(ilG_gui_frame *self, ilG_gui_coord coord);
enum ilG_gui_inputaction ilG_gui_click(ilG_gui_frame *top, int x, int y, int button);
void ilG_gui_hover(ilG_gui_frame *top, int x, int y);
void ilG_gui_draw(ilG_gui_frame *top);
void ilG_gui_addChild(ilG_gui_frame *parent, ilG_gui_frame *child);

]]

base.wrap "il.graphics.gui.frame" {
    struct = "ilG_gui_frame";
    filler = function(self, col)
        local v = ffi.new("float[4]")
        if type(col) == "table" then
            v[0] = col[1]
            v[1] = col[2]
            v[2] = col[3]
            v[3] = col[4]
        end
        return modules.graphics.ilG_gui_frame_filler(self, v)
    end;
    image = modules.graphics.ilG_gui_frame_image;
    click = modules.graphics.ilG_gui_click;
    label = function(self, layout, col, opts)
        local tex = texture()
        tex:setContext(self.context)
        tex:setName("Text Label")
        tex:fromimage("color0", layout:render(col, opts))
        self:image(tex, 1)
        local x, y = layout:getExtents()
        self:setSize(x,y)
    end;
    setPosition = function(self, x, y, xp, yp)
        xp = xp or 0
        yp = yp or 0
        local oldx, oldy = self.rect.a.x, self.rect.a.y
        local oldxp, oldyp = self.rect.a.xp, self.rect.a.yp
        self.rect.a.x = x
        self.rect.a.y = y
        self.rect.a.xp = xp
        self.rect.a.yp = yp
        self.rect.b.x = self.rect.b.x + x - oldx
        self.rect.b.y = self.rect.b.y + y - oldy
        self.rect.b.xp = self.rect.b.xp + xp - oldxp
        self.rect.b.yp = self.rect.b.yp + yp - oldyp
    end;
    setSize = function(self, w, h, wp, hp)
        wp = wp or 0
        hp = hp or 0
        self.rect.b.x = self.rect.a.x + w
        self.rect.b.y = self.rect.a.y + h
        self.rect.b.xp = self.rect.a.xp + wp
        self.rect.b.yp = self.rect.a.yp + hp
    end;
    addChild = modules.graphics.ilG_gui_addChild;
    abs = modules.graphics.ilG_gui_frame_abs;
    contains = modules.graphics.ilG_gui_frame_contains;
}

return modules.graphics.ilG_gui_frame_type;

File: src/graphics/gui/frame.c
#include "frame.h"

il_type ilG_gui_frame_type = {
    .typeclasses = NULL,
    .storage = NULL,
    .constructor = NULL,
    .destructor = NULL,
    .copy = NULL,
    .name = "il.graphics.gui.frame",
    .size = sizeof(ilG_gui_frame),
    .registry = NULL,
    .parent = NULL
};

static ilG_gui_coord computecoordabs(ilG_gui_coord coord, ilG_gui_rect parent)
{
    ilG_gui_coord size, res = {0,0,0,0};
    size.x = parent.b.x - parent.a.x;
    size.y = parent.b.y - parent.a.y;
    res.x = coord.x + size.x * coord.xp + parent.a.x;
    res.y = coord.y + size.y * coord.yp + parent.a.y;
    return res;
}

static ilG_gui_rect computerectabs(ilG_gui_rect rect, ilG_gui_rect parent)
{
    ilG_gui_rect res;
    res.a = computecoordabs(rect.a, parent);
    res.b = computecoordabs(rect.b, parent);
    return res;
}

ilG_gui_rect ilG_gui_frame_abs(ilG_gui_frame *self)
{
    if (self->parent) {
        return computerectabs(self->rect, ilG_gui_frame_abs(self->parent));
    } else {
        // all we can do is ignore the percentage field and treat the size as absolute
        ilG_gui_rect rect = {{0,0,0,0},{0,0,0,0}};
        rect.a.x = self->rect.a.x;
        rect.a.y = self->rect.a.y;
        rect.b.x = self->rect.b.x;
        rect.b.y = self->rect.b.y;
        return rect;
    }
}

int ilG_gui_frame_contains(ilG_gui_frame *self, ilG_gui_coord coord)
{
    ilG_gui_rect rect = ilG_gui_frame_abs(self);
    /*ilG_gui_coord size;
    size.x = rect.b.x - rect.a.x;
    size.y = rect.b.y - rect.a.y;*/
    coord = computecoordabs(coord, rect);
    return coord.x >= rect.a.x && coord.y >= rect.a.y &&
           coord.x <  rect.b.x && coord.y <  rect.b.y;
}

enum ilG_gui_inputaction ilG_gui_click(ilG_gui_frame *top, int x, int y, int button)
{
    ilG_gui_coord mouse = {x,y, 0.f, 0.f};
    if (!ilG_gui_frame_contains(top, mouse)) {
        return ILG_GUI_UNHANDLED;
    }
    enum ilG_gui_inputaction res = ILG_GUI_UNHANDLED;
    if (top->click) {
        res = top->click(top, x, y, button);
    }
    if (res == ILG_GUI_OVERRIDE) {
        return res;
    }
    ilG_gui_frame *cur = top;
    size_t i;
    for (i = 0; i < cur->children.length; i++) {
        enum ilG_gui_inputaction res2 = ilG_gui_click(cur->children.data[i], x, y, button);
        if (res2 > res) {
            res = res2;
        }
        if (res == ILG_GUI_OVERRIDE) {
            break;
        }
    }
    return res;
}

void ilG_gui_hover(ilG_gui_frame *top, int x, int y)
{
    (void)top, (void)x, (void)y;
}

static void draw(ilG_gui_frame *top, ilG_gui_rect rect)
{
    if (top->draw) {
        top->draw(top, rect);
    }
    size_t i;
    for (i = 0; i < top->children.length; i++) {
        draw(top->children.data[i], computerectabs(top->children.data[i]->rect, rect));
    }
}

void ilG_gui_draw(ilG_gui_frame *top)
{
    ilG_gui_rect rect = ilG_gui_frame_abs(top);
    draw(top, rect);
}

void ilG_gui_addChild(ilG_gui_frame *parent, ilG_gui_frame *child)
{
    IL_APPEND(parent->children, child);
    child->parent = parent;
}

